src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+                    CheckForShutdown();
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when shutting down, this checks for when shutdown is complete (no more active requests) and does actual disposal.
+        /// </summary>
+        /// <remarks>Requires <see cref="SyncObj"/> to be locked.</remarks>
+        private void CheckForShutdown()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObj));
+            Debug.Assert(ShuttingDown);
+
+            if (_activeRequests.Count != 0)
+            {
+                return;
+            }
+
+            if (_clientControl != null)
+            {
+                _clientControl.Dispose();
+                _clientControl = null;
+            }
+
+            if (_connection != null)
+            {
+                _connection.Close((long)Http3ErrorCode.NoError);
+                _connection.Dispose();
+                _connection = null;
+            }
+        }
+
+        public override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            Task waitTask = WaitForAvailableRequestStreamAsync(cancellationToken);
+
+            return waitTask.IsCompletedSuccessfully
+                ? SendWithoutWaitingAsync(request, cancellationToken)
+                : WaitThenSendAsync(waitTask, request, cancellationToken);
+        }
+
+        private async Task<HttpResponseMessage> WaitThenSendAsync(Task taskToWaitFor, HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            await taskToWaitFor.ConfigureAwait(false);
+            return await SendWithoutWaitingAsync(request, cancellationToken).ConfigureAwait(false);
+        }
+
+        private Task<HttpResponseMessage> SendWithoutWaitingAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            QuicStream quicStream = null;
+            Http3RequestStream stream;
+
+            try
+            {
+                // TODO: do less work in this lock.
+                lock (SyncObj)
+                {
+                    if (_connection == null)
+                    {
+                        goto retryRequest;
+                    }
+
+                    quicStream = _connection.OpenBidirectionalStream();
+                    if (_lastProcessedStreamId != -1 && quicStream.StreamId > _lastProcessedStreamId)
+                    {
+                        goto retryRequest;
+                    }
+
+                    stream = new Http3RequestStream(request, this, quicStream);
+                    _activeRequests.Add(quicStream.StreamId, stream);
+                }
+
+                return stream.SendAsync(cancellationToken);
+            }
+            catch (QuicResetException ex)
+            {
+                // This will happen if we aborted _connection somewhere.
+                Abort(ex);
+            }
+
+        retryRequest:
```

Comment:
nit: scrap the goto 😄

Reply:
why?

Reply:
It's more of a personal opinion. I think it generally makes code less readable and usually can be fixed via refactoring. We avoid them in AspNetCore, not sure if it varies in the runtime repo.

Reply:
I'll take another look to see if they can reasonably be removed without a size/perf penalty.

Reply:
> Does the goto mess up the lock statement?

C# doesn't let you create irreducible control-flow graphs with goto; so no. (i.e. you can't jump into a different scope).

It outputs a `leave.s` il instruction inside a lock, executing the finally first, undoing the lock

> The leave.s instruction is similar to the br instruction, but it can be used to exit a try, filter, or catch block whereas the ordinary branch instructions can only be used in such a block to > transfer control within it. The leave.s instruction empties the evaluation stack and ensures that the appropriate surrounding finally blocks are executed.

**tl;dr** C# won't compile if your goto use is unsound (unlike other languages)

Reply:
Aside, the C# compiler on the other hand... does create irreducible control-flow graphs in il, in certain circumstances  https://github.com/dotnet/roslyn/issues/39814

Reply:
Leaving this as a post-merge TODO. I'm not worried about the goto specifically but I'm going to go over the entire locking strategy, which was made a bit coarse to start with, and goto might naturally go away as part of that -- either way, once I get to this method I'll see if it's reasonable to get rid of it.